### ISSUE TITLE
"lakefs setup" should succeed if no changes needed to reach its end-goal

### ISSUE THREAD

**Issue by arielshaqed** [treeverse/lakeFS#2939]
When running `lakefs setup`, succeed (exit code 0) whenever no changes would be required to reach the expected goal of a working lakeFS.  (Currently it will fail if it has already run, or if the requested user already exists.)

This makes `lakefs setup` (usefully) _idempotent_: for instance, it allows (safe) automation of `lakefs setup` as part of setting up a lakeFS installation.

**Comment by ozkatz** [comment]
@arielshaqed can you elaborate on what changes are checked? existence of lakeFS DB schema? what happens if the lakefs binary is a newer version than the one that was previously provisioned? would it exit with >0 if there are unapplied migrations?

**Comment by itaiad200** [comment]
Our actions (lakectl + api) are not idempotent, why should this one be any different? I can argue the same about `lakectl superuser`, or even the API's `createBranch` for example. 

A safe automation can be accomplished by parsing the response. The 409 response that we return if the installation was already initialized, can we use that instead?

**Comment by arielshaqed** [comment]
> @arielshaqed can you elaborate on what changes are checked?

Good questions!

> existence of lakeFS DB schema?

Yes.

> what happens if the lakefs binary is a newer version than the one that was previously provisioned? would it exit with >0 if there are unapplied migrations?

You're right, I would expect it to `migrate up`.  Note that presumably automated deploy systems already do this for lakeFS.  The Helm chart [migrates up](https://github.com/treeverse/charts/blob/09de24319cc84e3adccc7e0cdf65dc9b995cb1e9/charts/lakefs/templates/job.yaml#L30) as [part of helm chart upgrade](treeverse/charts#33).  This is the only time a lakeFS upgrade can occur.  (However this Helm chart only gives instructions on how to set up, because it has no other way of giving the operator the initial credentials; see these [displayed NOTES](https://github.com/treeverse/charts/blob/4d8bd70cf2953d683410bbb32eddc10cb9834e86/charts/lakefs/templates/NOTES.txt#L3-L15).)

The plan is to make setup more explicit.

**Comment by arielshaqed** [comment]
> Our actions (lakectl + api) are not idempotent, why should this one be any different? I can argue the same about `lakectl superuser`, or even the API's `createBranch` for example.

I believe we _will_ get there.

> A safe automation can be accomplished by parsing the response. The 409 response that we return if the installation was already initialized, can we use that instead?

There is (currently) no safe way to parse the response.  If we must parse, I would be happier if we added a machine-readable output format to this (and potentially all) commands.  However that would still mean shell-scripting in init containers, or having to deploy a separate "setupper" image.

I would be happy to hide this behaviour behind a flag.  A big trend in setup systems is declarative configuration, and being able to run a command with an "I want the configuration to be like it would be after this command on a clean repo" flag should make it easier to write such declarations.

To be explicit: I am looking for something which can easily be run in a K8s init container (and [docker-compose now has init containers](https://stackoverflow.com/questions/70322031/does-docker-compose-support-init-container ) too).  Ideally it would require minimal work to create images -- so bonus points if I don't need a shell, and don't have to write my own code and build my own setup container.

**Comment by arielshaqed** [comment]
Thinking about it some more, my response to @ozkatz' questions was feature creep: I _grew_ the scope.  Instead, I should have outlined the gaps _and why I think **they are not important** for this issue_.  We can handle the gaps separately, as we grow more experience.

*tl;dr:* This feature remains useful if we define that a second `lakefs setup` fails after a lakeFS upgrade which changes the DB schema.

The remainder of this comment explains why -- so if you're already convinced, you might skip it. :-)

I would like to try to do that now (hopefully not too late).  I believe that this issue is perfectly useful, usable, and consistent with a smaller surface area than my original response.  (We might still end up doing everything in that response -- we can just be iterative about it and **not commit** right now to what we _might_ do in future.)

Where I'm coming from in algebra of combinators: The key point is that computer systems are (necessarily) [partial function][partial-function]s, so _idempotence only applies to successful executions_.  Call a combinator (function) F **idempotent** if for any state X, F(F(X))=F(X) _when all executions of F succeed_.  The last part is important: the system is allowed to return an error if some other function G was executed "concurrently" between the 2 executions of F.  (One example of such a combinator G might be "destroy S3 in us-east-1 with a well-aimed meteorite strike".  In such a case the outer call to F(F(X)) will fail.  A similar example might be "change the IAM role used by F to disallow writing to S3".)

Enough theory.

> > @arielshaqed can you elaborate on what changes are checked?
> 
> Good questions!

The answers are determined by which interfering operations G's we wish to handle and which we prefer to defer to operator intervention.

> 
> > existence of lakeFS DB schema?
> 
> Yes.

This still stands, and is the reason for this feature request.

> 
> > what happens if the lakefs binary is a newer version than the one that was previously provisioned? would it exit with >0 if there are unapplied migrations?

Originally (below) I expected it to migrate up and succeed.  G here is "upgrade deployed version of lakeFS".  While it is possible to migrate up (indeed I believe that it should), this is an unrelated feature.  `setup` F would still be idempotent if an upgrade G occurs and the second `setup` call fails.

Given that _the current state_ is for `lakefs run` to fail without an explicit `lakefs migrate up`, it is reasonable for `lakefs setup` to fail.  This just makes these 2 sequences behave the same -- both exit with error:

| | |
|---|---|
| `lakefs setup` | `lakefs setup` |
| _upgrade lakeFS_ | _upgrade lakeFS_ |
| `lakefs run` (fails) | `lakefs setup` (fails) |
|  | `lakefs run` (never tried, previous step failed |

So please ignore the remainder: it belongs to a separate discussion about how to handle upgrades in  managed environment.  We might have this discussion, but not now and _certainly_ not here.


[partial-function]: https://en.wikipedia.org/wiki/Partial_function

**Comment by arielshaqed** [comment]
The plot gets weirder: In fact we already have a *wider* behaviour that works.  From reading the code, running `lakefs setup` when already set up [succeeds silently, `Setup is already complete`](https://github.com/treeverse/lakeFS/blob/73ddc4d3e6de259e5c2cf7a6518834e5407d8b79/cmd/lakefs/cmd/setup.go#L67).

This is _not_ idempotent: after the sequence of commands
```sh
lakefs setup --user-name admin
lakefs setup --user-name control
```
where both commands succeeded, we have a system with a root user names `admin`.

So this issue has suddenly twisted to _disallowing_ some of this behaviour.  Specifically a second `lakefs setup` should succeed _only_ if no changes needed to reach its end-goal.  If the username (or the CLI-specified access key ID and/or secret key) have changed, it should fail.
